app/models/task/overdue_user_notification_list.rb

```diff
@@ -26,40 +22,55 @@ def initialize(month:, year:, output: STDOUT, logger: Rails.logger)
     def generate
       logger.info "Generating late contacts for #{year}, #{month}"
 
-      output.puts(CSV.generate_line(HEADER))
-      user_framework_lates.find_each do |user_framework|
-        output.puts(
-          CSV.generate_line(
-            [
-              user_framework.user_name,
-              user_framework.user_email,
-              user_framework.framework_short_name
-            ]
-          )
-        )
+      output.puts(CSV.generate_line(header))
+
+      suppliers.each do |supplier|
+        supplier.active_users.each do |user|
+          output.puts csv_line_for(user, supplier)
+        end
       end
     end
 
     private
 
-    ##
-    # NB should become its own class if this becomes an
-    # admin front end confection. This class should be concerned
-    # with generating CSV, not getting the records
-    def user_framework_lates
-      Task
-        .incomplete
-        .where(period_month: month, period_year: year)
-        .select('
-          tasks.id,
-          frameworks.short_name AS framework_short_name,
-          users.name AS user_name,
-          users.email AS user_email
-        ')
-        .joins(
-          :framework,
-          supplier: :active_users
-        )
+    def csv_line_for(user, supplier)
+      CSV.generate_line(
+        [user.email, due_date, user.name, supplier.name, reporting_month] + late_task_framework_presence(supplier)
+      )
+    end
+
+    def header
+      HEADER + frameworks_header
+    end
+
+    def reporting_month
+      @reporting_month ||= [Date::MONTHNAMES[month], year].join(' ')
+    end
+
+    def due_date
+      @due_date ||= ReportingPeriod.new(year, month).due_date.to_s(:day_month_year)
+    end
+
+    def frameworks
+      @frameworks ||= Framework.order(:short_name)
+    end
+
+    def frameworks_header
+      frameworks.map(&:short_name)
+    end
+
+    def late_task_framework_presence(supplier)
```

Comment:
Can this get memoized per supplier somehow? At the moment it gets re-run for every user within a supplier. Probably not worth worrying about though..?

Reply:
that's a good catch, and definitely a worthwhile optimisation, but I'm going to skip it for now as I'm eager to get this merged and deployed today so it can be run in the morning.